repo2docker/utils.py

```diff
@@ -69,6 +70,22 @@ def chdir(path):
     finally:
         os.chdir(old_dir)
 
+@contextmanager
+def open_utf8convert_read(path):
+    with open(path, "rb") as f:
+        file_to_encode = f.read()
+
+    encoding_detection_result = chardet.detect(file_to_encode)
+    if not "utf-8" in encoding_detection_result:
+        with open(path, "wb") as f:
+            f.write(file_to_encode.decode(encoding_detection_result["encoding"]).encode("utf-8"))
```

Comment:
I don't think we should overwrite the original file. If I was a user who carefully chose the encoding (for what ever reason) I'd be annoyed at repo2docker for overwriting it with some other encoding.

In (new-ish?) versions of Python we can pass the encoding of the file to `open`: https://docs.python.org/3/library/functions.html#open maybe we can ask chardet for the name of the encoding and pass it to `open()`. Or if this feature is too new (we support Python 3.5 I think) then maybe `chardet` has a `open()` that supports passing it an encoding?

Reply:
According to the docs we can open a file specifying its encoding in Python 3.5 also, yay :confetti_ball:

Reply:
Just checked we require Python 3.5+ so good to go!